joomla/joomla-cms
# Normalise JInstaller type adapters storeExtension methods and rollback on package install failure

Pull Request for several Issues.

### Summary of Changes

This PR makes a normalizes of JInstaller type adapters storeExtension methods, and also makes all extensions rollback when installing a package with several extensions.

Also normalizes install/update storeExtensions methods of all type adapters (except language that does not seems to use it). Probably, reading the old code, there are other issues that were solved also.

### Testing Instructions

This need to be tested carefully and with all install types: file, package, library, template, module, component and plugin.

A first test: Code review.

A second test:
Pre-requiste: Use this broken weblinks package (i made an error on the last extension manifest xml file so the package install fails). [pkg_weblinks-broken.zip](https://github.com/joomla/joomla-cms/files/671895/pkg_weblinks-broken.zip)

- Without patch install the broken package, notice it fails, but when you go to extension manage some extensions of the package are installed, also notice the module is installed in modules.
- Now manualy remove the leftover extensions trough Extension manage
- Now, with patch applied install the broken package, notice it fails, but when you go to extension manage NO extensions of the package are installed, also notice the module is NOT installed in modules.

More tests:
**This needs to be tested with all type of extensions: on install and on update**.

### Documentation Changes Required

None.

### Notes

- The component storeExtension method add some hardcoded stuff reggarding to delete exisiting (that the other types don't have), not sure if needed - removed it, if needed will add it back.
- Some extension types don't process the discover_install, i didn't touch that part.
